docs/source/upgrade.rst

```diff
@@ -0,0 +1,759 @@
+Instead, the :meth:`Extension.initialize <training.Extension.initialize>` method is added.
+This method is called by :meth:`Trainer.run <training.Trainer.run>` before entering the training loop.
+
+In Chainer v1, the extension is just called before entering the training loop when ``invoke_before_training`` is ``True``.
+**You need to fix the extension and override** :meth:`~training.Extension.initialize` **if you have custom extensions.**
+If you are using the :func:`~training.make_extension` decorator, you can set the ``initialize`` function by passing ``initializer`` to :func:`~training.make_extension`.
+
+.. _upgrade-dump-graph-only-once:
+
+The dump_graph extension dumps the valid graph only at its first invocation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In Chainer v2, the :func:`~training.extensions.dump_graph` extension dumps the valid computational graph only at its first invocation.
+**If you want to dump the graph more than once, you have to fix the code.**
+The easiest fix is setting the ``chainer.config.keep_graph_on_report`` flag to ``True``.
+*Note that this fix will cancel the improvement on the memory consumption made in Chainer v2.*
+More memory-efficient fix is to dump the graph without using an extension, e.g. by customizing the loss function or the updater.
+
+Here is the background of this change.
+In Chainer v2, :ref:`the Reporter copies reported variables with purging the computational graph by default. <upgrade-reporter-purge-variable>`
+On the other hand, the :func:`~training.extensions.dump_graph` extension requires the computational graph reachable from the reported variable.
+In order to make the graph available, the :func:`~training.extensions.dump_graph` extension turns on the ``chainer.config.keep_graph_on_report`` flag at its initializer (i.e., it turns on the graph before entering the training loop).
+Since we also wanted to achieve the memory efficiency, the :func:`~training.extensions.dump_graph` extension **turns off the flag after dumping the graph at its first invocation** (strictly speaking, it recovers the original value).
+As a result, the computational graph is not available from the second invocation.
+
+Since the :func:`~training.extensions.dump_graph` recovers the original flag value at its invocation, you can keep the graph dumped more than once by changing the original flag value.
+
+
+Reporter
+--------
+
+.. _upgrade-reporter-purge-variable:
+
+When a variable is reported, the variable is copied with the graph purged
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In Chainer v2, when a :class:`Variable` object is reported using :func:`report` function (or directly using :class:`Reporter`), a copy of the variable is made without preserving the computational graph.
+**If your code depends on the reachability of the computational graph from the reported variable, you have to update your code.**
```

Comment:
IMO, it is difficult for users to conclude with confident that their code is not broken by this change. Of course it is not possible to cover all code that will be broken, but it could be helpful to give an example that breaks this change.